core/domain/exp_domain.py

```diff
@@ -1346,11 +1346,10 @@ def validate(self, exp_param_specs_dict):
         for hint in self.hints:
             hint.validate()
 
-        if self.hints:
-            if self.solution:
-                self.solution.validate(self.id)
+        if self.hints and self.solution:
+            self.solution.validate(self.id)
 
-        elif self.solution:
+        if not self.hints and self.solution:
```

Comment:
Just to make it a bit easier for devs, I recommend writing this as `if self.solution and not self.hints`, so that it's super clear what the 'not' refers to.